### PR TITLE
std.string.index_unittest_examples

### PR DESCRIPTION
**Pull request by burner**
* added unittest doc examples for indexOf lastIndexOf

### CODE REVIEW

File: std/string.d [hunk 1]
```diff
@@ -484,6 +484,20 @@ ptrdiff_t indexOf(Range)(Range s, in dchar c,
     return -1;
 }
 
+///
+unittest
+{
+    string s = "Hello World";
+    auto i = indexOf(s, 'W');
+    assert(i == 6);
```
Comment (JackStouffer) [inline std/string.d]: Why not combine these lines to  `assert(s.indexOf('W') == 6);` or even  `assert("Hello World".indexOf('W') == 6);` if you want to get really compact? Same applies to the others lines in this.

Comment (burner) [inline std/string.d]: string s is used to mitigate copy paste erros.

The other part is personal style, I guess.

Comment (JackStouffer) [inline std/string.d]: IMO, UFCS makes examples easier to read, but it's not really that important.

**Comment by JackStouffer** [comment]
LGTM

**Comment by JakobOvrum** [comment]
These two functions already have really noisy documentation with all the duplication. This just exacerbates that issue. We should consolidate the docs for the various overloads.

**Comment by burner** [comment]
I thought that is accomplished by the "Jump to" links

**Comment by JakobOvrum** [comment]
Thought what was accomplished? AFAICS, the "Jump to" links at the top of an overload set only help linking to specific overloads, which is not the issue here. The issue is that each overload contains duplicate information, which really hurts readability and maintainability too (updating one of them requires updating all of them).

**Comment by burner** [comment]
I disagree, The current state allows the reader to see the parameter and the description for overload of indexOf in a structured way. If all overloads where merged into single documentation block, the distinguishing has to been textually. IMO that would be harder to read and to maintain. Additionally, these new examples, associated with a specific overload, make the docs very fast to read.

**Comment by andralex** [comment]
I do agree we must, as a general rule, document all overloads of a function together. It's the best way to organize documentation, and leads to good links. If overloads of a function are insufficiently related, they shouldn't be overloaded.

@burner, would you want to to work on that within this PR? The examples you added would be integrated with the others for the respective overloads.

**Comment by burner** [comment]
I'll work on that. Shouldn't be that difficult.

File: std/string.d [hunk 2]
```diff
@@ -500,6 +541,24 @@ ptrdiff_t indexOf(Range)(auto ref Range s, in dchar c,
     return indexOf!(StringTypeOf!Range)(s, c, cs);
 }
 
+///
```
Comment (MartinNowak) [inline std/string.d]: Surprised to see that having the documented unittests after an undocumented overload works.

Comment (burner) [inline std/string.d]: WAT, There is at least one error on my part, but hey magic.

**Comment by burner** [comment]
another try

File: std/string.d [hunk 3]
```diff
@@ -500,6 +559,13 @@ ptrdiff_t indexOf(Range)(auto ref Range s, in dchar c,
     return indexOf!(StringTypeOf!Range)(s, c, cs);
 }
 
+ptrdiff_t indexOf(Range)(auto ref Range s, in dchar c, in size_t startIdx,
```
Comment (andralex) [inline std/string.d]: So this is undocummented?

Comment (burner) [inline std/string.d]: Yes,

1. auto ref Range is only there to fix some rangification regressions.
2. all indexOf functions pretty much have the interface and are now also documented the same (compare cybershadow doc)

**Comment by quickfur** [comment]
ping @andralex 

Are @burner 's comments acceptable? Can we merge this yet? Let's get things moving forward.

File: std/string.d [hunk 4]
```diff
@@ -359,11 +359,17 @@ alias CaseSensitive = Flag!"caseSensitive";
...
 
     Returns:
-        the index of the first occurrence of $(D c) in $(D s). If $(D c)
+        the index of the first occurrence of $(D c) in $(D s) with
+        respect to the start index $(D startIdx). If $(D c)
         is not found, then $(D -1) is returned.
+        If $(D c) is found the value of the returned index is at least
+        $(D startIdx).
+        If the sequence starting at $(D startIdx) does not represent a well
+        formed codepoint, then a $(XREF utf,UTFException) may be thrown.
```
Comment (JakobOvrum) [inline std/string.d]: *May* be thrown? Not *is* thrown? It probably also belongs in the `Thrown:` section

**Comment by JakobOvrum** [comment]
Otherwise LGTM

File: std/string.d (hunk 4 above)
Comment (burner) [inline std/string.d]: Yes it is may be, it depends on the slice of the utf character  and the codepath taking.

Comment (JakobOvrum) [inline std/string.d]: Roger

Comment (burner) [inline std/string.d]: but I put it in a Throws: section

File: std/string.d [hunk 5]
```diff
@@ -1003,6 +1035,37 @@ ptrdiff_t lastIndexOf(Char)(const(Char)[] s, in dchar c,
     return -1;
 }
 
+/// Ditto
+ptrdiff_t lastIndexOf(Char)(const(Char)[] s, in dchar c, in size_t startIdx,
+        in CaseSensitive cs = CaseSensitive.yes) @safe pure
+    if (isSomeChar!Char)
+{
+    if (startIdx <= s.length)
```
Comment (andralex) [inline std/string.d]: can use `<` here

File: std/string.d [hunk 6]
```diff
@@ -1193,6 +1236,37 @@ ptrdiff_t lastIndexOf(Char1, Char2)(const(Char1)[] s, const(Char2)[] sub,
     return -1;
 }
 
+/// Ditto
+ptrdiff_t lastIndexOf(Char1, Char2)(const(Char1)[] s, const(Char2)[] sub,
+        in size_t startIdx, in CaseSensitive cs = CaseSensitive.yes) @safe pure
```
Comment (andralex) [inline std/string.d]: What's this supposed to if both s and sub are empty? Could you please add a unittest for that?

File: std/string.d (hunk 5 above)
Comment (burner) [inline std/string.d]: isn't slice top exclusive aka [) and doesn't that mean it has to be <=

```D
string a = "a";
assert(a[0 .. a.length] == "a");
```

File: std/string.d (hunk 6 above)
Comment (burner) [inline std/string.d]: 0e1afe82 introduced a bug. all (last,any)indexOf should return -1 if the needle can't be found and you can't find nothing.

**Comment by burner** [comment]
@andralex update

**Comment by andralex** [comment]
thx!

**Comment by burner** [comment]
thanks

**Pull request merged**